numpy/core/src/private/ufunc_override.c

```diff
@@ -149,31 +164,58 @@ PyUFunc_WithOverride(PyObject *args, PyObject *kwds,
                 Py_DECREF(method);
                 goto fail;
             }
-            if (with_override != NULL) {
-                Py_INCREF(obj);
-                with_override[num_override_args] = obj;
-            }
-            if (methods != NULL) {
-                methods[num_override_args] = method;
-            }
-            else {
-                Py_DECREF(method);
-            }
+            Py_INCREF(obj);
+            with_override[num_override_args] = obj;
+            methods[num_override_args] = method;
             ++num_override_args;
         }
     }
     return num_override_args;
 
 fail:
-    if (methods != NULL) {
-        for (i = 0; i < num_override_args; i++) {
-            Py_DECREF(methods[i]);
+    for (i = 0; i < num_override_args; i++) {
+        Py_DECREF(with_override[i]);
+        Py_DECREF(methods[i]);
+    }
+    return -1;
+}
+
+/*
+ * Check whether any of a set of input and output args have a non-default
+ *  `__array_ufunc__` method. Return 1 if so, 0 if not.
+ *
+ * Does not check that args is a tuple and kwds either NULL or a dict.
+ */
+NPY_NO_EXPORT int
+PyUFunc_HasOverride(PyObject *args, PyObject *kwds)
+{
+    int i;
+    int nin, nout;
+    PyObject *out_kwd_obj;
+
+    /* check inputs */
+    nin = PyTuple_GET_SIZE(args);
+    for (i = 0; i < nin; ++i) {
+        if (has_non_default_array_ufunc(PyTuple_GET_ITEM(args, i))) {
+            return 1;
         }
     }
-    if (with_override != NULL) {
-        for (i = 0; i < num_override_args; i++) {
-            Py_DECREF(with_override[i]);
+    /* check outputs, if any */
+    if (kwds == NULL) {
+        return 0;
+    }
+    out_kwd_obj = PyDict_GetItemString(kwds, "out");
```

Comment:
Not a huge fan of parsing the out argument in a third place. Can we use `ufunc_full_args` somehow?

Reply:
I'd like to leave that to a later PR (in #11372, it ends being done at the very start of `ufunc_generic_call`,  which I think is most logical, since every part below it needs it).

Reply:
Fine by me - my objection was to introducing one more parser. I'm happy now that the number of parsers in this file hasn't increased :)